Add custom filter listener.

demiankatz opened the pull request:
This PR makes it possible to represent complex filters using simpler placeholders in the URL. It also creates the ability to have "inverted" filters, where filters are applied to Solr UNLESS a specific filter value is applied in the URL. Both of these features are particularly useful in combination with checkbox facets.

It's entirely possible that there is room for improvement in the configuration layout and comments; I'm interested in feedback on these (and anything else).

TODO:
- [x] Write tests
- [x] Discuss configuration format

demiankatz committed:
Add custom filter listener.

demiankatz committed:
Fix styles.

demiankatz committed:
php-cs-fixer.

demiankatz commented on module/VuFind/src/VuFind/Search/Solr/CustomFilterListener.php:
```diff
@@ -0,0 +1,155 @@
+     *
+     * @var string
+     */
+    protected $filterParam = 'fq';
```
I put this into a property because I think we could theoretically move most of the logic to an abstract base class and then have separate implementations for other search backends like Summon.

EreMaijala replied:
Ok. This reminds me.. we really should try to move more of the backend-specific stuff to VuFindSearch, I think.

demiankatz replied:
I think the original intention of putting the listeners in VuFind instead of VuFindSearch was that the listeners were implementing functionality specific to VuFind rather than generic to backends... but that may no longer be true of all of the listeners, or we may no longer care about that distinction. Certainly happy to revisit. :-)

EreMaijala replied:
What I meant is that e.g. 'fq' is pretty specific to how the url query syntax with Solr currently works. There's been at least some level of agreement that the legacy facets api that we use should be deprecated with Solr 9, which probably means that in the long run we need to switch to the JSON APIs, for all queries. Then the 'fq' in VuFind classes becomes a burden. So, while I think it's fine that a listener is specific to a backend, I'd like the interface between VuFind and VuFindSearch to be more backend-agnostic. It's for sure not this pull request that needs to address it. :)

demiankatz replied:
That's a good point, though I don't think there's an easy solution to this problem. The original design decision was to have the contents of the ParamBag passed to the search backend be backend-specific, and to have the Params classes take care of the primary mapping of VuFind parameters to backend parameters. Of course, custom listeners live somewhat outside of this piece, but in the end, I think the problem here is deciding what needs to be given a VuFind-level abstraction, and where we put the mapping logic. Maybe it would be helpful to make facets more abstract so that more code can be shared across backends, but given the inconsistencies in implementations on different platforms (particular with regard to things like AND/OR/NOT facets), that may not be practical. And even if we do decide to make facets uniform, there are going to be other backend-specific features that perhaps do not make sense to develop abstractions for, because the use cases are too narrow. Anyway, you rightly point out that we need to work on the JSON facets implementation, and perhaps that work will inspire our next step forward -- that certainly seems like a good context in which to continue this conversation.

demiankatz committed:
Add test.

demiankatz committed:
Add tests; simplify code.

demiankatz commented:
@EreMaijala, I believe this is ready for review now. There are a lot of different ways we could structure the configuration, and I'm open to alternative ideas if you don't like what I came up with here. I tried to separate the behavior of the listener from the structure of the configuration to keep things flexible.

EreMaijala commented on config/vufind/facets.ini:
```diff
@@ -304,3 +304,31 @@ visual_facets = "callnumber-first,topic_facet"
+; UNLESS the custom filters are applied by the user. This is useful in
+; combination with checkbox facets if you want to exclude values by default
+; and provide a checkbox that allows them to be included.
+;inverted_filters[include_books] = '-format:"Book"'
```
Missing newline at EOF

demiankatz replied:
Fixed.

EreMaijala commented on config/vufind/facets.ini:
```diff
@@ -304,3 +304,31 @@ visual_facets = "callnumber-first,topic_facet"
+; below, and your custom filter field was the default of __custom__, you
+; would add &filter[]=__custom__:include_books to the URL in order to apply
+; the inverted filter.
+custom_filter_field = "__custom__"
```
Would it be possible to use "vufind" as the default custom filter field? It would look nicer. Also, it would be nice to include an example of how to set up a checkbox filter to use the field, and in its configuration 'vufind' would look nicer and be easier to get right than double underscore.

demiankatz replied:
Agreed. I have made the change (and added the example... though the comment on the example is perhaps too short because I didn't want to go multi-line; open to improvements if you have any suggestions).

demiankatz replied:
...actually, the comment was bothering me, so I revised it. Hopefully better now!

EreMaijala reviewed (changes requested):
Looks good to me, just some minor niggles.

demiankatz committed:
Merge branch 'dev' into custom-filter-listener

demiankatz committed:
Address code review comments.

demiankatz committed:
Reformat comments.

demiankatz committed:
Avoid notices.

demiankatz commented:
Thanks, @EreMaijala -- I've addressed your review comments and also fixed a small bug that was turning up when running the full test suite (missing configuration was causing a notice when building the SolrAuth backend).

EreMaijala reviewed (approved):
Looks good to me!
